### PR TITLE
[IOTDB-1541] Change sequence of wal and memtable in insert

### PR DESCRIPTION
**Pull request by HeimingZ**
When inserting data, IoTDB first inserts data into memtable and then writes wal. If writing wal occurs error, the data in memtable should rollback, but IoTDB doesn't implement this. This problem can be fixed by changing the sequence of wal and memtable in insert.

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java [hunk 1]
```diff
@@ -213,8 +213,6 @@ public void insert(InsertRowPlan insertRowPlan) throws WriteProcessException {
       checkMemCostAndAddToTspInfo(insertRowPlan);
     }
 
-    workMemTable.insert(insertRowPlan);
-
     if (IoTDBDescriptor.getInstance().getConfig().isEnableWal()) {
       try {
         getLogNode().write(insertRowPlan);
```
Comment (HTHou) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java:218]: Move the wal to the front of the memCost check?

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java [hunk 2]
```diff
@@ -272,12 +272,12 @@ public void insertTablet(
       throw new WriteProcessException(e);
     }
     try {
-      workMemTable.insertTablet(insertTabletPlan, start, end);
       if (IoTDBDescriptor.getInstance().getConfig().isEnableWal()) {
         insertTabletPlan.setStart(start);
         insertTabletPlan.setEnd(end);
         getLogNode().write(insertTabletPlan);
       }
+      workMemTable.insertTablet(insertTabletPlan, start, end);
```
Comment (HTHou) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java:280]: Same as above

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java (hunk 1 above)
Comment (HeimingZ) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java:214]: fixed

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java (hunk 2 above)
Comment (HeimingZ) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java]: fixed

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java [hunk 3]
```diff
@@ -261,6 +261,19 @@ public void insertTablet(
...
 
+    try {
+      if (IoTDBDescriptor.getInstance().getConfig().isEnableWal()) {
+        insertTabletPlan.setStart(start);
+        insertTabletPlan.setEnd(end);
+        getLogNode().write(insertTabletPlan);
+      }
+    } catch (IOException e) {
+      for (int i = start; i < end; i++) {
+        results[i] = RpcUtils.getStatus(TSStatusCode.WRITE_PROCESS_REJECT, e.getMessage());
```
Comment (HTHou) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java:272]:
```suggestion
        results[i] = RpcUtils.getStatus(TSStatusCode.INTERNAL_SERVER_ERROR, e.getMessage());
```

File: server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java [hunk 4]
```diff
@@ -261,6 +261,19 @@ public void insertTablet(
...
+      }
+      throw new WriteProcessException(e);
```
Comment (HTHou) [inline server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java:274]:
```suggestion
    } catch (Exception e) {
      for (int i = start; i < end; i++) {
        results[i] = RpcUtils.getStatus(TSStatusCode.INTERNAL_SERVER_ERROR, e.getMessage());
      }
      throw new WriteProcessException(e);
```

Comment (HeimingZ) [reply server/src/main/java/org/apache/iotdb/db/engine/storagegroup/TsFileProcessor.java]: fixed

**Review by qiaojialin — APPROVED** [review]

**Pull request merged**